xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -305,6 +310,34 @@ func (cib *clusterImplBalancer) UpdateState(state balancer.State) {
 	cib.pickerUpdateCh.Put(state)
 }
 
+func (cib *clusterImplBalancer) getClusterName() (string, bool) {
+	v := cib.clusterName.Load()
+	vv, ok := v.(string)
+	return vv, ok
+}
+
+func (cib *clusterImplBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if clusterName, ok := cib.getClusterName(); ok {
+		newAddrs := make([]resolver.Address, len(addrs))
+		for i, addr := range addrs {
+			newAddrs[i] = xdsinternal.SetHandshakeClusterName(addr, clusterName)
+		}
+		addrs = newAddrs
+	}
+	return cib.ClientConn.NewSubConn(addrs, opts)
```

Comment:
If `getClusterName` returned false for its second return parameter, does it make sense to create this subConn without the clusterName attribute? And if so, why and when does it make sense to do so?

Reply:
If ok is false, the atomic.Value is never set. But the value is set in the same method that this balancer gets the addresses and passes them down.

So in practice, this would happen if the child policy creates some SubConn before it gets any addresses (impossible for balancers like roundrobin, but can happen for some sophisticated ones, like xds).

We can
1. delay the NewSubConn(), and wait for at least one ClientConn update so it has clusterName
1. return an error
1. create a SubConn without attributes

1 is an overkill.
2 is probably OK
3 is better IMO because the attributes only affect cases like direct path. Most creds doesn't read the attributes. And in direct path, the connection will fail at handshake.

Let me know what you think.

Reply:
Sounds good.

Reply:
This check is no longer needed with the mutex approach.